Migration of international/localization/currencies page form and some exchange rate actions

tomas862 opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This branch is created from https://github.com/PrestaShop/PrestaShop/pull/10774. It will have all missing pieces in order to have fully functional currency page (Form and live exchange rates update )
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10190
| How to test?  | See below

## How to test

This PR migrates 2 things:
- on the "Localization > Currencies" listing, the "Exchange rates" block
- when you click on "edit" or "add new currency" on "Localization > Currencies" listing you access the add/edit form, it has been migrated
URL of the listing `/admin-dev/index.php/improve/international/currencies/`
URL of the edit form for currency 1 : `/admin-dev/index.php/improve/international/currencies/1/edit`

⚠️ You need to rebuild assets before testing

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/11894)
<!-- Reviewable:end -->

tomas862 committed:
displays empty form page

tomas862 committed:
displays empty form fields

tomas862 committed:
displays currencies retrieved from cldr package

tomas862 committed:
adds default cldr currency as well as displays cldr currencies data to the selector

tomas862 committed:
adds currency settings

tomas862 committed:
creates currency form handler skeleton

tomas862 committed:
adds status switch

tomas862 committed:
adds shop association dropdown

tomas862 committed:
adds basic currency creation implementation

tomas862 committed:
adds currency update command and its implementation

tomas862 committed:
adds currency retrieval query bus command

tomas862 committed:
adds shop association creation in separate abstract class

tomas862 committed:
adds multishop column update functionality

tomas862 committed:
fixes multistore problems, adds some todos

tomas862 committed:
adds currency iso code as separate value object

tomas862 committed:
adds currency exchange rate as an object with its own validations

tomas862 committed:
adds modified commands usage implementation also in update currency command handler

tomas862 committed:
adds assertion which checks existing currency on create currency action

tomas862 committed:
adds another assertion which validates from existing currency modification when new iso code is being changed

tomas862 committed:
adds exception handling, adds some fixes regarding existing currency validation

tomas862 committed:
adds constraint validations

tomas862 committed:
adds demoRestricted access checking

tomas862 committed:
displays exchange rate settings panel

tomas862 committed:
adds witdh of panel to be in half of the page

tomas862 committed:
adds hint messages for live exchange rates

tomas862 committed:
adds validation for cron jobs module installed or uninstalled status checking

tomas862 committed:
adds update using an ajax call

tomas862 committed:
fixes restriction checking logic, adds some todoes

tomas862 committed:
adds dynamic text update on switch change

tomas862 committed:
adds dynamic switch text update

tomas862 committed:
creates command for updating live exchange rates, adds basic checking of installed cronjob module

tomas862 committed:
receives cron url

tomas862 committed:
creates the cron for live currency exchange rates and also adds existing setting validation

tomas862 committed:
adds exception handling for live exchange rates update

tomas862 committed:
adds button for manual exchange rates update

tomas862 committed:
adds manual exchange rate functionality update command

tomas862 committed:
adds missing comments

tomas862 committed:
adds some unit tests for currency is code and exchange rate

tomas862 committed:
adds that the currency tab will point to new currency controller, fixed some spelling mistakes

tomas862 committed:
adds assertion which prevents from disabling default currency on edit form action

tomas862 committed:
removes currency controller templates as well as legacy admin currency controller

tomas862 committed:
applies cs fixer

tomas862 committed:
removes some vulnerable comment from the exception message

tomas862 commented:
This PR is ready for the review!
Don't forget to rebuilt assets

The page can be accessed normally via Improve -> International -> Localization -> Currencies tab.

@matks I guess lets close this https://github.com/PrestaShop/PrestaShop/pull/10774 first and then I rebase this PR as well :slightly_smiling_face:

sarjon commented on admin-dev/themes/new-theme/js/pages/currency/CurrencyLiveExchangeRate.js:
```diff
@@ -0,0 +1,72 @@
+/**
```
@matks i see js files being named using `camelCase.js` or `dash-separated.js`, should we establish some naming convetion here as well?

matks replied:
Yes indeed

sarjon commented on admin-dev/themes/new-theme/js/pages/currency/CurrencyLiveExchangeRate.js:
```diff
@@ -0,0 +1,72 @@
+        showSuccessMessage(response.message);
+        this._changeTextByCurrentSwitchValue($switch.val());
+      }
+    ).fail((response) => {
```
i suggest using `catch` as it is part of `Promise` API instead of jQuery's `fail()`

tomas862 replied:
I was unable to catch an error with `catch` - seems like due to I send different status code on error for instance, 401, but I'm not really sure about it :thinking:

sarjon commented on admin-dev/themes/new-theme/js/pages/currency/index.js:
```diff
@@ -0,0 +1,50 @@
+  choiceTree.enableAutoCheckChildren();
+
+  const currencyLiveExchangeRate = new CurrencyLiveExchangeRate();
+  currencyLiveExchangeRate.initEvents();
```
why not `initEvents()` in constructor?

tomas862 replied:
otherwise I would need to call `new CurrencyLiveExchangeRate()` only - in such case it looks like an unused expression.

sarjon replied:
well, i dont think that judging on "how it looks" is correct. In addition, "initEvents()" does not add much "usefulness", does it?

In this case calling `new CurrencyLiveExchangeRate()` is like enabling some javascript component for page. In my opinion, you could improve it by renaming it and adding some comments, like:

```javascript
// index.js

// allows scheduling cronjob via ajax for currency exchange rates update from "Live exchange rates" form
new ExchangeRatesUpdateScheduler();
```

sarjon commented on src/Adapter/Currency/CommandHandler/CreateCurrencyHandler.php:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @internal
+ */
+final class CreateCurrencyHandler extends AbstractObjectModelLegacyHandler implements CreateCurrencyHandlerInterface
```
@matks same here, we have mix between `Add*Command` and `Create*Command`, i think we should settle with using only one verb either `Add` or `Create`, wdyt?

matks replied:
Actually both are rights if properly used: `Create` : you create something from scratch. `Add` : you add something to a list of items.

Most of usecases should `Create` as in CRUD. But `Add` might be used sometimes.

sarjon commented on src/Adapter/Currency/CommandHandler/CreateCurrencyHandler.php:
```diff
@@ -0,0 +1,100 @@
+use PrestaShopException;
+
+/**
+ * Class CreateCurrencyHandler is responsible for adding new currency.
```
why "for **adding** new currency" when commands says **create**? :smile:

sarjon commented on src/Adapter/Currency/CommandHandler/CreateCurrencyHandler.php:
```diff
@@ -0,0 +1,100 @@
+     */
+    public function handle(CreateCurrencyCommand $command)
+    {
+        $this->assertCurrencyCodeAlreadyExist($command->getIsoCode()->getValue());
```
from assertion name it seems that it's ok when currency with given ISO code already exists, but actually its not.

what i think it should say is `assertCurrencyWithIsoCodeDoesNotExist($newIsoCode)`, this means that you just want to make sure that currency with your new iso code does not exist before you create new one, otherwise it will fail at asserting and never reach the code below.

sarjon commented on src/Adapter/Currency/CommandHandler/CreateCurrencyHandler.php:
```diff
@@ -0,0 +1,100 @@
+                ];
+            }
+
+            $this->updateMultiStoreColumns($entity, $columnsToUpdate);
```
its not clear what this code between lines 68-75 does. what i suggest is extracting it into separate method `private function associateConvertionRate($convertionRate, array $shopIds)` or similar, wdyt?

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateCurrencyHandler.php:
```diff
@@ -0,0 +1,181 @@
+ *
+ * @internal
+ */
+final class UpdateCurrencyHandler extends AbstractObjectModelLegacyHandler implements UpdateCurrencyHandlerInterface
```
same here. most of handlers are named as `Edit*Handler` since its mroe related to Back Office action names.

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateCurrencyHandler.php:
```diff
@@ -0,0 +1,181 @@
+                );
+            }
+
+            $this->assertNewIsoCodeAlreadyExists(
```
again here, i dont think you want to assert that currency with iso code exists. :wink:

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateCurrencyHandler.php:
```diff
@@ -0,0 +1,181 @@
+                $entity->iso_code,
+                $command->getIsoCode()->getValue()
+            );
+            $this->assertIsDefaultCurrencyEnabled($command->getCurrencyId()->getValue(), $command->isEnabled());
```
is it really what you want to assert? wouldnt it make more sense to `assertDefaultCurrencyIsNotBeingDisabled(Currency $currency)`, wdyt?

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateCurrencyHandler.php:
```diff
@@ -0,0 +1,181 @@
+                    'conversion_rate' => $entity->conversion_rate,
+                ];
+            }
+            $this->updateMultiStoreColumns($entity, $columnsToUpdate);
```
same here, you can extract this piece of code into separate method, maybe even abstract class and use in both `AddCurrencyHandler` and `EditCurrencyHandler`, wdyt?

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateExchangeRatesHandler.php:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @internal
+ */
+final class UpdateExchangeRatesHandler implements UpdateExchangeRatesHandlerInterface
```
i think it would be more appropriate to call it `RefreshExchangeRatesHandler` and `RefreshExchangeRatesCommand` as it is what documentation says, wdyt?

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateLiveExchangeRatesHandler.php:
```diff
@@ -0,0 +1,252 @@
+    {
+        if (!$this->isCronJobModuleInstalled) {
+            throw new DisabledLiveExchangeRatesException(
+                'Live exchange rates feature cannot be modified due to cronjob module is uninstalled'
```
you dont know if module was installed in the first place. :p

so maybe write this:
```diff
- Live exchange rates feature cannot be modified due to cronjob module is uninstalled
+ Live exchange rates feature cannot be modified because "cronjob" module is not installed
```

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateLiveExchangeRatesHandler.php:
```diff
@@ -0,0 +1,252 @@
+            );
+        }
+
+        $this->configuration->restrictUpdatesTo($this->contextShop);
```
i would be very careful about using this, if you were in "All shops" contexts and use `Configuration` service again, would it work as expected?

tomas862 replied:
This part is tricky:

I left exactly how it was on legacy currencies controller - it was restricted to certain shop context only. And yes, on "All shops context" it works in unexpected way.

However, the stuff it generates is the url which calls `cron_currency_rates.php` file which iterates over all shops anyway :o 

I guess it was made to be that way that the cron would be executed only on one shop :thinking: 

maybe here I need to pass `contextShopIds` which will be able to create cron urls for all shops - but this will result in same file call from all shop contexts.

Not sure how to do here. Any thoughts @matks , @sarjon  ?

matks replied:
OK, we are dealing with a wild usecase here 😄 

I think we're going to leave it the way it is right now because modifying this behavior requires a lot of rework.

What should be done in this rework:
- it should be required to pass `cron_currency_rates.php` a shop ID so the script knows which shop rates need to be updated
- each cron entry should call the script passing it's own shop ID so a merchant can really control which shop gets updated and how often

This is what makes sense to me 🤔 

If it looks right to you, can you create an issue so we dont forget to do this (I'll put it in the Migration kanban) and add a `@todo` statement in the code ? This way we won't forget (when we have migrated all SF pages, we'll scan for all `@todo` statements and this should indicate what's left to do).

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateLiveExchangeRatesHandler.php:
```diff
@@ -0,0 +1,252 @@
+            }
+
+            if (!$cronId && $command->isExchangeRateEnabled() && false === $this->createCronJob($cronUrl)) {
+                throw new CannotCreateLiveExchangeUpdateCronTaskException(
```
`CannotCreateLiveExchangeUpdateCronTaskException` name is too complex imo. maybe you could have one `ScheduleExhangeRatesUpdateException` with different error codes? wdyt?

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateLiveExchangeRatesHandler.php:
```diff
@@ -0,0 +1,252 @@
+        $cronId = $this->configuration->get('PS_ACTIVE_CRONJOB_EXCHANGE_RATE');
+
+        try {
+            $cronUrl = $this->getCronUrl();
```
you only use `$coronUrl` after `if` statement below, so you can define it after that.

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateLiveExchangeRatesHandler.php:
```diff
@@ -0,0 +1,252 @@
+ *
+ * @internal
+ */
+final class UpdateLiveExchangeRatesHandler implements UpdateLiveExchangeRatesHandlerInterface
```
for me handler/command name is a bit misleading, it gives impression that `UpdateLiveExchangeRatesHandler` updates exchange rates _now_, but what it actually does is _schedules_ cron job for that.
how about naming it `ScheduleExchangeRatesUpdateHandler` or something similar? @matks wdyt?

matks replied:
Same feeling as you: the name is a bit misleading

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateLiveExchangeRatesHandler.php:
```diff
@@ -0,0 +1,252 @@
+    public function handle(UpdateLiveExchangeRatesCommand $command)
+    {
+        if (!$this->isCronJobModuleInstalled) {
+            throw new DisabledLiveExchangeRatesException(
```
i think exception is wrong, `ScheduleExchangeRateUpdateException` with code `ScheduleExchangeRateUpdateException::CRON_TASK_MANAGER_MODULE_NOT_INSTALLED` would be better imo, @matks maybe you have other suggestions?

sarjon commented on src/Adapter/Currency/CommandHandler/UpdateLiveExchangeRatesHandler.php:
```diff
@@ -0,0 +1,252 @@
+            }
+
+            if ($cronId) {
+                $this->validateCronJob($cronId);
```
maybe `resetCronConfiguration()` would be a better name since it resets configuration if needed?

sarjon commented on src/Adapter/Currency/QueryHandler/GetCurrencyForFormEditingHandler.php:
```diff
@@ -0,0 +1,97 @@
+ *
+ * @internal
+ */
+final class GetCurrencyForFormEditingHandler implements GetCurrencyForFormEditingHandlerInterface
```
im wondering if you have to be that specific with name "GetCurrencyFor**Form**EditingHandler". :thinking: 
from my understanding, our domain code does not care from what UI (whether its html form, CLI or rest API) it's being edited. @matks wdyt, do we need to keep "Form" in command / handlers names?

tomas862 replied:
I guess you are right here :+1:

sarjon commented on src/Adapter/Currency/QueryHandler/GetCurrencyForFormEditingHandler.php:
```diff
@@ -0,0 +1,97 @@
+            );
+        }
+
+        try {
```
seems that this `try/catch` is useless, isnt it?

sarjon commented on src/Adapter/Currency/QueryHandler/GetCurrencyForFormEditingHandler.php:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        try {
+            $result = new CurrencyForFormEditing(
```
you can return it directly. in addition, i'd suggest you naming it `EditableCurrency`.

sarjon commented on src/Adapter/Domain/AbstractObjectModelLegacyHandler.php:
```diff
@@ -0,0 +1,135 @@
+use Shop;
+
+/**
+ * Class AbstractObjectModelLegacyHandler is responsible for executing legacy code which is common for legacy ObjectModel.
```
```diff
- Class AbstractObjectModelLegacyHandler is responsible for executing legacy code which is common for legacy ObjectModel.
+ Encapsulates common behavior for legacy object models
```

matks replied:
Fun thing: Sarunas PR has exactly this need https://github.com/PrestaShop/PrestaShop/pull/11885#discussion-diff-243558128R159

sarjon commented on src/Adapter/Module/ModuleValidator.php:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @internal
+ */
+class ModuleValidator
```
it does not seem that you have a good reason to introduce this service. 1. it dies not have interface 2. its just wrapper around `ModuleManager`.

in my opinion, this service should be deleted and you better relay on already existing service to check if module is installed. @matks wdyt?

sarjon commented on src/Adapter/Tools.php:
```diff
@@ -204,4 +204,14 @@ public function isModRewriteActive()
+     *
+     * @return string
+     */
+    public function getShopProtocol()
```
i dont see value in these kind of methods, it same legacy code. :/ 

i would better use `Tools::getShopProtocol()` directly in my adapter handlers but avoid introducing new API without interface.

sarjon commented on src/Core/Domain/Currency/Command/CreateCurrencyCommand.php:
```diff
@@ -0,0 +1,112 @@
+     * @param CurrencyExchangeRate $exchangeRate
+     * @param bool $isEnabled
+     */
+    public function __construct(CurrencyIsoCode $isoCode, CurrencyExchangeRate $exchangeRate, $isEnabled)
```
can you accept `scalar` values and create VOs inside constructor?

tomas862 replied:
you mean in the constructor to pass only $isoCode as string and assign it to VO in the constructor itself?

sarjon replied:
yup.

sarjon commented on src/Core/Domain/Currency/Command/DeleteCurrencyCommand.php:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * @param CurrencyId $currencyId
+     */
+    public function __construct(CurrencyId $currencyId)
```
same here and for other commands, could you accept scalar values and create VOs inside constructor?

sarjon commented on src/Core/Domain/Currency/Command/UpdateExchangeRatesCommand.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Class UpdateExchangeRatesCommand
+ */
+class UpdateExchangeRatesCommand
```
shouldnt it extend some parent?

tomas862 replied:
command doesn't have any info to send here - I don't see any point of extending here something :thinking:

sarjon commented on src/Core/Domain/Currency/Command/UpdateLiveExchangeRatesCommand.php:
```diff
@@ -0,0 +1,78 @@
+     */
+    public function __construct($isExchangeRateEnabled)
+    {
+        $this->assertType($isExchangeRateEnabled);
```
`assertIsBool` ?

sarjon commented on src/Core/Domain/Currency/Exception/CannotRefreshExchangeRatesException.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Class CannotRefreshExchangeRatesException
+ */
+class CannotRefreshExchangeRatesException extends CurrencyException
```
maybe `RefreshExchangeRatesException` would be enough?

sarjon commented on src/Core/Domain/Currency/ValueObject/CurrencyExchangeRate.php:
```diff
@@ -0,0 +1,87 @@
+    public function __construct($exchangeRate)
+    {
+        $this->assertExchangeRate($exchangeRate);
+        $this->exchangeRate = (float) $exchangeRate;
```
do not cast it, better assert that is it `float` instead of `numeric`. from what i see it could be `assertIsFloatOrZero` since these are 2 of 3 constraints for exchange rate, right?

sarjon commented on src/Core/Domain/Currency/ValueObject/CurrencyExchangeRate.php:
```diff
@@ -0,0 +1,87 @@
+                    'Exchange rate is not valid. Expected to be numeric but got "%s"',
+                    var_export($exchangeRate, true)
+                ),
+                CurrencyConstraintException::INVALID_EXCHANGE_RATE_TYPE
```
i think you can keep `CurrencyConstraintException::INVALID_EXCHANGE_RATE` here as well, unless you care that invalid value is due wrong type or because its less than 0, but imo its not the case. wdyt?

sarjon commented on src/Core/Domain/Currency/ValueObject/CurrencyId.php:
```diff
@@ -0,0 +1,64 @@
+            );
+        }
+
+        $this->currencyId = (int) $currencyId;
```
do not cast it here, instead check if its `is_int`.

sarjon commented on src/Core/Domain/Currency/ValueObject/CurrencyIsoCode.php:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Class IsoCode
+ */
+class CurrencyIsoCode
```
would be nice to know if this code is numeric or alpha code. you could just add docs about it with examples. :wink: 

also, you can remove `Currency` prefix from name imo.

sarjon commented on src/Core/Domain/Currency/ValueObject/CurrencyIsoCode.php:
```diff
@@ -0,0 +1,89 @@
+                    'Iso code is not valid. Excpected to be string but got "%s"',
+                    var_export($isoCode, true)
+                ),
+                CurrencyConstraintException::INVALID_ISO_CODE_TYPE
```
same here, do you really need separate error code when invalid value is due to invalid type?

sarjon commented on src/Core/Domain/Currency/ValueObject/CurrencyIsoCode.php:
```diff
@@ -0,0 +1,89 @@
+        if (!is_string($isoCode)) {
+            throw new CurrencyConstraintException(
+                sprintf(
+                    'Iso code is not valid. Excpected to be string but got "%s"',
```
i think you dont need to put " around %s when using `var_export`.

sarjon commented on src/Core/Form/ChoiceProvider/CurrencyFromCldrChoiceProvider.php:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Class CurrencyFromCldrChoiceProvider
+ */
+final class CurrencyFromCldrChoiceProvider implements FormChoiceProviderInterface
```
name is confusing a bit, maybe `CurrencyNameByIsoCodeChoiceProvider` would be better? :thinking:

sarjon commented on src/Core/Domain/Currency/ValueObject/CurrencyExchangeRate.php:
```diff
@@ -0,0 +1,87 @@
+/**
+ * Class CurrencyLiveExchangeRate
+ */
+class CurrencyExchangeRate
```
no need to prefix with `Currency`, namespace gives it a context.

sarjon commented on src/Core/Form/IdentifiableObject/DataProvider/CurrencyFormDataProvider.php:
```diff
@@ -0,0 +1,85 @@
+    public function getDefaultData()
+    {
+        return [
+            'exchange_rate' => 1,
```
what is 1? how about storing is as constant in `ExchangeRate` VO as `ExchangeRate::DEFAULT_RATE`?

sarjon commented on src/Core/Grid/Definition/Factory/CurrencyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,191 @@
+    {
+        return (new FilterCollection())
+            ->add((new Filter('iso_code', TextType::class))
+            ->setTypeOptions([
```
some cs fixer is missing. :p

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -0,0 +1,505 @@
+        }
+
+        return $this->render('@PrestaShop/Admin/Improve/International/Currency/create.html.twig', [
+            'is_shop_feature_enabled' => $multiStoreFeature->isUsed(),
```
twig vars should be camelCase. :wink:

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -0,0 +1,505 @@
+
+        try {
+            if ($currencyForm->isSubmitted() && $this->isDemoModeEnabled()) {
+                $this->addFlash('error', $this->getDemoModeErrorMessage());
```
if i remember correctly, demo mode is handled in form handling system, so you should not worry too much about it.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -0,0 +1,505 @@
+                        'Admin.Notifications.Success'
+                    ),
+                ];
+                $statusCode = 200;
```
there is const for that in `Response` class. :wink:

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -0,0 +1,505 @@
+            'status' => false,
+            'message' => $this->trans('Unexpected error occurred.', 'Admin.Notifications.Error'),
+        ];
+        $statusCode = Response::HTTP_INTERNAL_SERVER_ERROR;
```
i'd prefer 400 (bad request) instead of 500 as default status code.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -0,0 +1,505 @@
+            return $exceptionDictionary[$exceptionClass][$exceptionCode];
+        }
+
+        return $this->trans('Unexpected error occurred.', 'Admin.Notifications.Error');
```
you can fallback to `$this->getFallbackErrorMessage()`

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -0,0 +1,505 @@
+            return $exceptionTypeDictionary[$exceptionType];
+        }
+
+        return $this->trans('Unexpected error occurred.', 'Admin.Notifications.Error');
```
you can fallback to $this->getFallbackErrorMessage()

sarjon commented on src/PrestaShopBundle/Form/Admin/Improve/International/Currencies/CurrencyFormDataProvider.php:
```diff
@@ -0,0 +1,68 @@
+     */
+    public function setData(array $data)
+    {
+        // TODO: Implement setData() method.
```
todo?

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/improve/international/currencies.yml:
```diff
@@ -0,0 +1,63 @@
+        _legacy_controller: AdminCurrencies
+
+admin_currencies_create:
+    path: /create
```
```diff
- path: /create
+ path: /new
```

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/improve/international/currencies.yml:
```diff
@@ -0,0 +1,63 @@
+        _legacy_controller: AdminCurrencies
+
+admin_currencies_update_exchange_rates:
+    path: /update-exchange_rates
```
```diff
- path: /update-exchange_rates
+ path: /update-exchange-rates
```

sarjon commented on src/PrestaShopBundle/Resources/config/services/core/cldr.yml:
```diff
@@ -0,0 +1,8 @@
+    _defaults:
+        public: true
+
+    prestashop.core.cldr.repository:
```
repository is already registered under `prestashop.core.localization.cldr.repository` key, so you can use it.

or you can make alias since this one is using key that matches namesapce.

sarjon commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Currency/Blocks/form.html.twig:
```diff
@@ -0,0 +1,93 @@
+
+  </div>
+
+  {{ form_end(currencyForm) }}
```
dont forget `form_rest`

sarjon commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Currency/exchange_rates.html.twig:
```diff
@@ -0,0 +1,45 @@
+{{ form_start(exchangeRatesForm, {'action': path('admin_currencies_update_exchange_rates') }) }}
```
shouldnt it be under `/Blocks` as it is block of page?

sarjon reviewed (commented):
quite a few comments, but good job overall. :+1:

tomas862 committed:
renames some functions to have appropriate names

tomas862 committed:
creates new function to associatio conversion rate columns to make it easier to understand code part

tomas862 committed:
renames create handler to add handler

tomas862 committed:
renames update to edit currency

tomas862 committed:
creates an abstraction for Add and Edit currency handlers to move shared logic

tomas862 committed:
renames some more functions, exceptions, classes to have appropriate names

tomas862 committed:
renames javascript files

tomas862 committed:
renames exception, adds exception code for it

tomas862 committed:
renames some more handlers

LouiseBonnard commented on src/PrestaShopBundle/Form/Admin/Improve/International/Currencies/CurrencyType.php:
```diff
@@ -0,0 +1,102 @@
+                    ]),
+                    new GreaterThan([
+                        'value' => 0,
+                        'message' => $this->trans('This field is invalid', [], 'Admin.Notifications.Error'),
```
Could we add here an extra indication on how to make this field valid?

tomas862 replied:
sure I can - although there will be more new stuff to translate :)

LouiseBonnard replied:
Nice! Let's wait for @TristanLDD to come back and check with him the UX/UI friendliness of this new wording, aha! I suggest:

- Oops, this field is invalid - it must contain numeric values
- Sorry, invalid field, it must contain numeric values

LouiseBonnard replied:
Well, here is what we've agreed on:

- This field is invalid, it must contain numeric values

tomas862 replied:
@LouiseBonnard  just to clarify - even though it validates that number must be greater than 0 we should display message: "This field is invalid, it must contain numeric values" anyway?

LouiseBonnard commented on src/PrestaShopBundle/Form/Admin/Improve/International/Currencies/CurrencyType.php:
```diff
@@ -0,0 +1,102 @@
+                        'message' => $this->trans('This field is invalid', [], 'Admin.Notifications.Error'),
+                    ]),
+                ],
+                'invalid_message' => $this->trans('This field is invalid', [], 'Admin.Notifications.Error'),
```
Same here.

tomas862 replied:
it validates the input type - in this case it should be numeric. 

How the message should look like - maybe "This field is invalid - it must be numeric type". ? :thinking:

LouiseBonnard replied:
Well, here is what we've agreed on:

- This field is invalid, it must contain numeric values

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Currency/exchange_rates.html.twig:
```diff
@@ -0,0 +1,45 @@
+          </label>
+          <div class="col-sm">
+            <button class="btn btn-primary">
+              {{ 'Update'|trans({}, 'Admin.Global') }}
```
I think the word class here is the verb, not the noun - and in this case, I would rather localize this string in _Admin.Actions_. ;-)

LouiseBonnard commented on src/Core/Grid/Definition/Factory/CurrencyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,191 @@
+     */
+    protected function getName()
+    {
+        return $this->trans('Currencies', [], 'Admin.Navigation.Menu');
```
Can you confirm that this string is to be found within the menu of the back office?

tomas862 replied:
yes - checked with applying another language to it and it was translated :)

tomas862 committed:
renames query handler

tomas862 committed:
removes uselles try catch block

tomas862 committed:
checks cronjob module is installed condition in a different way

tomas862 committed:
removes some non required adapter functions

tomas862 committed:
renames value objects, exception messages, functions

tomas862 committed:
renames some value objects, fixes some code style

tomas862 committed:
renames twig variables to camelCase

tomas862 committed:
renames some classes in the test suit

tomas862 committed:
removes demo restricted checkings

tomas862 committed:
changed status codes

tomas862 committed:
uses fallback error message function to get the default message

tomas862 committed:
moves exchange rate live update to settings form submit handler

tomas862 committed:
removed one cldr service usage

tomas862 committed:
includes form_rest function

tomas862 committed:
moves exchange rate template to Block folder

tomas862 committed:
adds improved translations

tomas862 committed:
uses value objects inside the constructor

tomas862 committed:
removes type castings inside commands

tomas862 committed:
fixes tests, adds strict types validation for ExchangeRate and IsoCode VOs

tomas862 committed:
applies cs fixer

tomas862 committed:
passes array and not null as a multistore shop parameter, adds a logic which prevents removal of default currency in its shop context

tomas862 committed:
changes the translation message

tomas862 commented:
@LouiseBonnard  I added entirely new validation case and maybe I need to have a message for it.

![default_currency_removal](https://user-images.githubusercontent.com/17051250/50678382-20a4d680-1007-11e9-81d8-df4ac1ff586c.png)

It turns out that before it was possible to remove default currency from shop association. I added validation but not sure if I need a message for it.

I'm thinking I can add a messages something like this:

1. **%field% is required for shop %shop_name%** in Admin.Notifications.Error
2. **%currency% is the default for shop %shop_name% and cannot be removed**

first one might be applied for more cases while the second one is only for this specific case.

tomas862 committed:
prevents from validating shop associations when it is not in the shop context

tomas862 committed:
adds empty array checking before updating shop associations

tomas862 committed:
moves javascript events initialization to constructor

matks commented:
@tomas862 https://github.com/PrestaShop/PrestaShop/pull/10774 is merged 😉

LouiseBonnard commented:
@tomas862, thanks for pinging me on this. Since we are working on improving the error notifications with @TristanLDD, I would like to wait for his feedback before - and he will be back on Monday.

- _The field %field_name% is required for shop %shop_name%_ in Admin.Notifications.Error
- _%currency% is the default currency for shop %shop_name%, and therefore cannot be removed from shop association_ in Admin.International.Notification or Admin.Notifications.Error

tomas862 committed:
renames form handlers, builders

tomas862 committed:
removes Admin currencies controller from tests

matks commented on admin-dev/themes/new-theme/js/pages/currency/ExchangeRatesUpdateScheduler.js:
```diff
@@ -0,0 +1,76 @@
+   * @private
+   */
+  _initLiveExchangeRate(event) {
+    const $switch = $(event.currentTarget);
```
Why `$switch` as variable name ?

matks commented on src/Core/Domain/Currency/Command/ScheduleExchangeRatesUpdateCommand.php:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Class ScheduleExchangeRatesUpdateCommand
+ */
+class ScheduleExchangeRatesUpdateCommand
```
If it's used to enable or disable, maybe we can give it a name related to this matter ? Like ToggleSwitchOnOrOff, EnableOrDisable ..?

tomas862 replied:
it is used for setting up cronjob or disabling it - if it would be named like `ToggleExchangeRatesUpdateCommand` then it would be a bit misleading due to it does not instantly execute exchange rates update. And also, I have a similar Command which does that instantly :)

matks replied:
What about something like `ToggleExchangeRateAutomatizationCommand` ? It `toggles` the `exchange rates` `automatization process`.

matks commented on src/Adapter/Currency/CommandHandler/ScheduleExchangeRatesUpdateHandler.php:
```diff
@@ -0,0 +1,247 @@
+     *
+     * @throws CurrencyException
+     */
+    public function handle(ScheduleExchangeRatesUpdateCommand $command)
```
It took me a few minutes to understand what this function was doing 😅 

Idea in order to make it more readable:
```
public function handle(ScheduleExchangeRatesUpdateCommand $command)
{
    if ($command->isExchangeRateEnabled()) {
        $this->enableExchangeRate();
    else {
        $this->disableExchangeRate();
    }
}
```

matks replied:
And since this depends on `$cronId` result, I would even imagine a "matrix" of usecases:
- cron should be turned on
- cron should be turned off
- cron is already active
- no cron active

```
if ($cronShouldBeTurnedOff && $noCronRunning) {
    return;
elseif ($cronShouldBeTurnedOn && $cronIsRunning) {
    return;
else ($cronShouldBeTurnedOff && $cronIsRunning) {
    $this->turnCronOff();
else {
    $this->turnCronOn();
}
```

matks commented on src/Adapter/Currency/CommandHandler/ScheduleExchangeRatesUpdateHandler.php:
```diff
@@ -0,0 +1,247 @@
+        $cronId = $this->configuration->get('PS_ACTIVE_CRONJOB_EXCHANGE_RATE');
+
+        try {
+            if ($cronId && !$command->isExchangeRateEnabled()) {
```
In order to make the final goal more readable, what about:
```
$thereIsOneCronRunning = ($cronId !== 0):
if ($thereIsOneCronRunning && !$command->isExchangeRateEnabled()) {
```

matks commented on src/Core/Domain/Currency/Command/ScheduleExchangeRatesUpdateCommand.php:
```diff
@@ -0,0 +1,78 @@
+    /**
+     * @var bool
+     */
+    private $isExchangeRateEnabled;
```
What about `$shouldExchangeRateBeEnabled` ? Because this is what is intended, not what is real right now. Or `$exchangeRateStatus`

matks commented on src/Adapter/Currency/CommandHandler/ScheduleExchangeRatesUpdateHandler.php:
```diff
@@ -0,0 +1,247 @@
+     *
+     * @throws PrestaShopException
+     */
+    private function getCronUrl()
```
I'm worried that the day someone modifies the `cron_currency_rates.php` script or refactor it and provides it a clean URL such as `script/currency_rate/update` this dev will never think about updating this.

Ideas to improve:
- create a dedicated class (CurrencyRateScriptUrlProvider ?) in charge of generating the right URL
- add a function to `LinkCore` class for this usecase

tomas862 replied:
But I'm thinking it won't help to remember for someone if anything is added to the LinkCore or new class is being added like `CurrencyRateScriptUrlProvider` because:

If I start refactoring `cron_currency_rates.php` I know it is the file and somehow it should be called from somewhere so I search across whole project to find `cron_currency_rates.php` string
This will lead the way to the handler `ScheduleExchangeRatesUpdateHandler`

If I add To `LinkCore` new method my search method of this file usage will not change - I would do the same :thinking:

matks replied:
OK since this is the only place where it is used 👍 I agree
If there was more than 1 place where it would be used, it would be nice to have a method to call 😉

matks commented on src/Adapter/Currency/CommandHandler/ScheduleExchangeRatesUpdateHandler.php:
```diff
@@ -0,0 +1,247 @@
+     *
+     * @throws Exception
+     */
+    private function resetCronConfiguration($cronId)
```
The `reset` word might be misleading here. Without looking a the code, I thought that using `resetCronConfiguration()` would call `removeCronJob()` if eligible.

I'm thinking about something more like `fixConfigurationIfWrong()` ? 😅

matks commented on src/Core/Domain/Currency/Query/GetCurrencyForEditing.php:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * @param CurrencyId $currencyId
+     */
+    public function __construct(CurrencyId $currencyId)
```
Please accept and integer but cast it into a CurrencyId 😉

matks commented on src/Core/Form/IdentifiableObject/DataHandler/CurrencyFormDataHandler.php:
```diff
@@ -0,0 +1,90 @@
+            $data['active']
+        );
+
+        $command->setShopIds(!is_array($data['shop_association']) ? [] : $data['shop_association']);
```
```suggestion
        $command->setShopIds(is_array($data['shop_association']) ? $data['shop_association'] : []);
```
This way you remove the `!` 😄

matks commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -95,24 +106,45 @@ public function searchAction(Request $request)
+        }
+
+        return $this->render('@PrestaShop/Admin/Improve/International/Currency/create.html.twig', [
+            'isShopFeatureEnabled' => $multiStoreFeature->isUsed(),
```
👍

matks commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -205,6 +253,135 @@ public function toggleStatusAction($currencyId)
+     */
+    public function updateLiveExchangeRatesAction(Request $request)
+    {
+        if ($this->isDemoModeEnabled()) {
```
I guess this is because `@DemoRestricted` does not fit JSON responses. But maybe we can improve it so we can use here ? Add something like `@AjaxDemoRestricted` or `@JsonDemoRestriced`

matks commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -205,6 +253,135 @@ public function toggleStatusAction($currencyId)
+            );
+        }
+
+        $authLevel = $this->authorizationLevel($request->attributes->get('_legacy_controller'));
```
However I think you can use `@AdminSecurity` even for ajax routes. Maybe you did not want to do it because you did not want to redirect ?

Again, maybe there is room for improvement. If you do not want to make this PR bigger, tell me and we'll make it the subject for another PR 😄

tomas862 replied:
For @AjaxDemoRestricted and @AdminSecurity for ajax I though about we can do something like that - but it seems like totally different functionality so I would rather introduce it in another PR to keep things simple and it would be way easier for someone to review it ;) Should I add reminder for myself or maybe we should raise an issue with it for improvement? :thinking:

matks replied:
Done 😄 https://github.com/PrestaShop/PrestaShop/issues/12083

matks commented on src/PrestaShopBundle/Form/Admin/Improve/International/Currencies/CurrencyFormDataProvider.php:
```diff
@@ -0,0 +1,85 @@
+
+    /**
+     * @param CommandBusInterface $commandBus
+     * @param int $cronExchangeRate
```
On line 40 you say it's a boolean 🤔

matks commented on src/PrestaShopBundle/Form/Admin/Improve/International/Currencies/CurrencyFormDataProvider.php:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * @var bool
+     */
+    private $cronExchangeRate;
```
If it's a boolean, can we rename it with `is` as prefix ? Such as `$isCronExchangeActive`

matks commented on src/PrestaShopBundle/Resources/config/services/adapter/currency.yml:
```diff
@@ -17,3 +17,44 @@ services:
+          - '@=service("prestashop.adapter.legacy.context").getContext().shop'
+          - '@prestashop.adapter.shop.shop_url'
+          - '@translator'
+          - '@=service("prestashop.adapter.data_provider.module").isInstalled("cronjobs")'
```
I'm worried because we have hardcoded a module name here in the SF config.

There is definitely something wrong with this module that is finally very coupled with the Core 😅 but I would like to avoid hardcoded values into the Symfony configuration if possible. I'd prefer to pass the `prestashop.adapter.data_provider.module` service and have the hardcoded module name in the code.

tomas862 replied:
I'm using `@=service("prestashop.adapter.data_provider.module").isInstalled("cronjobs")` in two places - Core class `CurrencyExchangeRateType` and in Adapter `ScheduleExchangeRatesUpdateHandler`.

Unfortunately this class does not have any interface and it is used in Adapter. An Adapter usage in Core should be avoided at all costs as I learned :thinking: 

When we get rid of the module usage instead of passing `@=service("prestashop.adapter.data_provider.module").isInstalled("cronjobs")`  we can pass some kind of configuration which defines if this feature is on or off or we will remove it anyway and it will be just one line to remove when doing ctrl +F across whole project. If I use it differently in Adapter there is a greater change that somewhere this might be forgotten :smiley:

matks replied:
That makes sense 👍 let's leave it like this

matks commented on src/PrestaShopBundle/Resources/config/services/core/cldr.yml:
```diff
@@ -1,8 +0,0 @@
-    _defaults:
-        public: true
-
-    prestashop.core.cldr.repository:
```
We can't do that. It's a BC break if someone built a shop on PS 1.7.5 and started using this service. Do you wish to deprecate it ?

tomas862 replied:
It was introduced on October 4th, 2018 in this commit https://github.com/PrestaShop/PrestaShop/commit/99aec1594844fc6c94f69756a6296d8f615f4a7d.

I removed it due to @sarjon told that this service already exists in

`src/PrestaShopBundle/Resources/config/services/core/localization/cldr.yml` but now I see that it was introduced on October 10th, 2018 so @sarjon should remove hes  service and I should keep it :rofl: 

as I see the commit https://github.com/PrestaShop/PrestaShop/commit/99aec1594844fc6c94f69756a6296d8f615f4a7d. was merged only 5 days ago so its not in 1.7.5 release, right? :thinking:

matks replied:
OK I see it now.
You're right: we can delete it, but not here because I know at least one other PR that relies on one of these services.
Please use the service you want 😄 and I'll make a PR that deletes one of them and "migrate" all usages to the right one.

matks replied:
Issue created: https://github.com/PrestaShop/PrestaShop/issues/12118

matks commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Currency/Blocks/exchange_rates.html.twig:
```diff
@@ -0,0 +1,45 @@
+                  {{ 'Please install the %module_name% module before using this feature.'|trans({'%module_name%': 'cronjobs'}, 'Admin.International.Notification') }}
+                </small>
+              {% else %}
+                <small class="form-text js-exchange-rate-text-when-disabled {{ (exchangeRatesForm.live_exchange_rate.vars.value == 0) ? '' : 'd-none' }}">
```
Why not an `if` / `else` here ? Is it the legacy behavior ?

tomas862 replied:
I need both messages to be available, just to hide one and display another on different condition due to the on/off switch is being triggered via Ajax so I need a certain text to appear :slightly_smiling_face:

matks commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Currency/Blocks/form.html.twig:
```diff
@@ -0,0 +1,93 @@
+
+        {% if isShopFeatureEnabled %}
+          <div class="form-group row">
+            <label for="" class="form-control-label">
```
Empty `for` ? Maybe we can add hardcoded one for UX ?

matks commented on tests-legacy/PrestaShopBundle/Controller/ControllerTest.php:
```diff
@@ -97,7 +97,6 @@ public function getControllersClasses()
             array('AdminCarriersController'),
             array('AdminStatusesController'),
             array('AdminZonesController'),
-            array('AdminCurrenciesController'),
```
You can add the symfony test in https://github.com/PrestaShop/PrestaShop/blob/develop/tests-legacy/Integration/PrestaShopBundle/Controller/Admin/SurvivalTest.php 😉

matks commented on tests/Unit/Core/Domain/Currency/ValueObject/CurrencyExchangeRateTest.php:
```diff
@@ -0,0 +1,82 @@
+
+    public function getIncorrectExchangeRates()
+    {
+        return [
```
Well ... it seems a big array with not so much data in it ?

Do you have something you dont like about
```
[0],
[-1],
['0'],
['-1'],
['4.294.967.295,000'],
[null],
[[]],
```

matks reviewed (changes requested):
@tomas862 Review done 😄 whao so much work ! I've got a few questions / issues but overall good work

tomas862 committed:
renames switch variable to more precise one

tomas862 committed:
refactors function so its more clear what it really does

tomas862 committed:
passes scalar value for currency edit DTO

tomas862 committed:
switches is array comparison places

tomas862 committed:
renames some comments

tomas862 committed:
adds label for="" and id value, adds currency controller in symfony tests

tomas862 committed:
reduces some test case variables

tomas862 commented:
@matks - thanks for the Review ! :+1: , I applied fixes were I was able to and I left some questions ;)

Also to be able to finish with all translations I'm waiting for reply to the message in  https://github.com/PrestaShop/PrestaShop/pull/11894#issuecomment-451413611

tomas862 committed:
uses es lint to fix files syntax

TristanLDD commented:
@LouiseBonnard I'm ok with the wording ;)

LouiseBonnard commented:
@TristanLDD, thanks! @tomas862, I let you apply the following wording:

- _The field %field_name% is required for shop %shop_name%_ in _Admin.Notifications.Error_
- _%currency% is the default currency for shop %shop_name%, and therefore cannot be removed from shop association_ in _Admin.International.Notification_ or _Admin.Notifications.Error_

matks commented:
@tomas862 answered all questions 😉

tomas862 committed:
renames exchange rate scheduler

tomas862 committed:
handles currency cannot be removed from shop context exception

tomas862 committed:
applies php cs fixer

tomas862 commented:
@matks - PR is ready! ;)

tomas862 commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -125,18 +158,36 @@ public function createAction()
+            $this->addFlash('error', $this->handleException($exception));
+        }
+
+        return $this->render('@PrestaShop/Admin/Improve/International/Currency/create.html.twig', [
```
todo: edit.html.twig use here

matks replied:
Yes I agree 😄

tomas862 committed:
changes to another template instead using the same for both create and update actions

tomas862 committed:
applies php cs fixer

tomas862 committed:
fixes yml file validity

tomas862 committed:
adds some renaming, fixes fatal error in edit handler

tomas862 committed:
applies php cs fixer

tomas862 committed:
prevents from forever loop when permissions are not granted in currency page

matks commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -125,18 +155,36 @@ public function createAction()
+            $this->addFlash('error', $this->handleException($exception));
+        }
+
+        return $this->render('@PrestaShop/Admin/Improve/International/Currency/update.html.twig', [
```
`edit.html.twig` instead of `update.html.twig` 😉

matks commented on src/Core/Domain/Currency/Exception/CannotRemoveDefaultCurrencyFromShopAssociationException.php:
```diff
@@ -26,9 +26,51 @@
 
 namespace PrestaShop\PrestaShop\Core\Domain\Currency\Exception;
 
+use Throwable;
```
We have to support php5.6 so we cannot use `Throwable`

tomas862 replied:
evil phpStorm ide auto included this for me :imp: I'll fix that :+1:

PierreRambaud replied:
Come to Emacs my friend :trollface: Come to the dark side!

matks commented:
@tomas862 3 minor issues 😉

sarjon commented on admin-dev/themes/new-theme/js/pages/currency/ExchangeRatesUpdateScheduler.js:
```diff
@@ -0,0 +1,76 @@
+export default class ExchangeRatesUpdateScheduler {
+  constructor() {
+   this._initEvents();
+  }
```
you can return `{}` since this class does not have any "public" api.

sarjon commented on admin-dev/themes/new-theme/js/pages/currency/ExchangeRatesUpdateScheduler.js:
```diff
@@ -0,0 +1,76 @@
+
+const $ = window.$;
+
+export default class ExchangeRatesUpdateScheduler {
```
what does it do? (docs missing)

sarjon commented on src/Adapter/Currency/CommandHandler/AbstractCurrencyHandler.php:
```diff
@@ -0,0 +1,56 @@
+use PrestaShop\PrestaShop\Adapter\Domain\AbstractObjectModelHandler;
+
+/**
+ * Class AbstractCurrencyHandler is responsible for encapsulating common behavior for legacy object models.
```
```suggestion
 * Class AbstractCurrencyHandler is responsible for encapsulating common behavior for legacy currency object model.
```

sarjon commented on src/Adapter/Currency/CommandHandler/AddCurrencyHandler.php:
```diff
@@ -0,0 +1,91 @@
+     */
+    public function handle(AddCurrencyCommand $command)
+    {
+        $this->assertCurrencyWithIsoCodeDoesNotExist($command->getIsoCode()->getValue());
```
:+1:

sarjon commented on src/Adapter/Currency/CommandHandler/EditCurrencyHandler.php:
```diff
@@ -0,0 +1,233 @@
+     */
+    public function __construct(
+        $defaultCurrencyId,
+        $isMultiStoreFeature
```
```diff
- $isMultiStoreFeature
+ $isMultistoreFeatureUsed
```

sarjon commented on src/Adapter/Currency/CommandHandler/EditCurrencyHandler.php:
```diff
@@ -0,0 +1,233 @@
+                $command->getIsoCode()->getValue()
+            );
+            $this->assertDefaultCurrencyIsNotBeingDisabled($command->getCurrencyId()->getValue(), $command->isEnabled());
+            $this->assertDefaultCurrencyIsBeingRemovedFromShop(
```
```diff
- assertDefaultCurrencyIsBeingRemovedFromShop
+ assertDefaultCurrencyIsNotBeingRemovedFromShop
```

sarjon commented on src/Adapter/Currency/CommandHandler/EditCurrencyHandler.php:
```diff
@@ -0,0 +1,233 @@
+        $qb
+            ->select('id_currency')
+            ->from('currency')
+            ->where('id_currency !=' . $currencyId)
```
```suggestion
            ->where('id_currency !=' . (int) $currencyId)
```

just in case.

sarjon commented on src/Adapter/Currency/CommandHandler/EditCurrencyHandler.php:
```diff
@@ -0,0 +1,233 @@
+     *
+     * @throws CannotDisableDefaultCurrencyException
+     */
+    private function assertDefaultCurrencyIsNotBeingDisabled($currencyId, $isEnabled)
```
i think you can pass whole command instead of 2 args since you want to assert that command is not disabling default currency, wdyt?

sarjon commented on src/Adapter/Currency/CommandHandler/ToggleExchangeRateAutomatizationHandler.php:
```diff
@@ -0,0 +1,262 @@
+
+            if ($thereIsOneCronRunning && !$command->exchangeRateStatus()) {
+                $this->disableExchangeRatesScheduler($cronId);
+            }
```
i didnt check it, but does it make sense to go to each `if`? let's say 2nd if is `true`, do you still need to check if 3rd `if` is `true`? in other words, shouldnt it be `if`/`else if` ?

sarjon commented on src/Adapter/Currency/CommandHandler/ToggleExchangeRateAutomatizationHandler.php:
```diff
@@ -0,0 +1,262 @@
+        if (!$this->isCronJobModuleInstalled) {
+            throw new ScheduleExchangeRatesUpdateException(
+                'Live exchange rates feature cannot be modified because "cronjob" module is not installed',
+                ScheduleExchangeRatesUpdateException::CRON_TASK_MANAGER_MODULE_NOT_INSTALLED
```
i think you should either use `Schedule` or `Automate` (following your handler name).

sarjon commented on src/Core/Domain/Currency/Command/ToggleExchangeRateAutomatizationCommand.php:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Class ToggleExchangeRateAutomatizationCommand
+ */
+class ToggleExchangeRateAutomatizationCommand
```
can you describe what this command does?

sarjon commented on src/Core/Domain/Currency/DTO/EditableCurrency.php:
```diff
@@ -0,0 +1,99 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Currency\DTO;
```
it should no logner be part in `DTO` namespace but instead `QueryResult`.

sarjon commented on src/Core/Domain/Currency/DTO/EditableCurrency.php:
```diff
@@ -0,0 +1,99 @@
+     * @param bool $isEnabled
+     * @param int[] $associatedShopIds
+     */
+    public function __construct($isoCode, $exchangeRate, $isEnabled, array $associatedShopIds)
```
i think it would make sense to also return currency id?

sarjon commented on src/Core/Domain/Currency/Exception/CannotCreateCurrencyException.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Class CannotCreateCurrencyException
+ */
+class CannotCreateCurrencyException extends CurrencyException
```
@matks i see we are having a lot of mixing between add/create and edit/update in all places (routes, controllers, actions), do you think it's ok?

@tomas862 you dont need to change this, it's just something to think about for us. :)

matks replied:
@sarjon It's an area for improvement but it's not blocking 😃

sarjon commented on src/Core/Domain/Currency/Exception/CannotRemoveDefaultCurrencyFromShopAssociationException.php:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Class CannotRemoveDefaultCurrencyFromShopAssociationException
+ */
+class CannotRemoveDefaultCurrencyFromShopAssociationException extends CurrencyException
```
hmmm, maybe `DefaultCurrencyException` and code `CANNOT_REMOVE_DEFAULT_CURRENCY` would be better?

tomas862 replied:
well for this specific exception I need `$currencyName, $shopName` to pass as arguments so the new exception helps to avoid them in another exception where this arguments are not needed

sarjon commented on src/Core/Domain/Currency/Exception/ScheduleExchangeRatesUpdateException.php:
```diff
@@ -0,0 +1,36 @@
+namespace PrestaShop\PrestaShop\Core\Domain\Currency\Exception;
+
+/**
+ * Class CannotCreateLiveExchangeUpdateCronTask
```
does not match class name.

sarjon commented on src/Core/Domain/Currency/Exception/ScheduleExchangeRatesUpdateException.php:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Class CannotCreateLiveExchangeUpdateCronTask
+ */
+class ScheduleExchangeRatesUpdateException extends CurrencyException
```
is it `Schedule` or `Automate`? i see this everywhere.

sarjon commented on src/Core/Domain/Currency/ValueObject/ExchangeRate.php:
```diff
@@ -0,0 +1,79 @@
+     *
+     * @throws CurrencyConstraintException
+     */
+    private function assertIsFloatOrZero($exchangeRate)
```
cant it be integer?

sarjon commented on src/Core/Domain/Currency/ValueObject/IsoCode.php:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Class IsoCode
+ */
+class IsoCode
```
what kind of iso code it is (alpha or numeric)? you should either document it or name `AlphaIsoCode` imo.

sarjon commented on src/Core/Form/IdentifiableObject/DataHandler/CurrencyFormDataHandler.php:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Class CurrencyFormDataHandler
+ */
+final class CurrencyFormDataHandler implements FormDataHandlerInterface
```
check all your services to have proper documentation.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -95,24 +104,45 @@ public function searchAction(Request $request)
+        $currencyForm->handleRequest($request);
+
+        try {
+            if ($currencyForm->isSubmitted()) {
```
you dont need to check if its submited. :wink:

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -125,18 +155,36 @@ public function createAction()
+            $currencyForm = $this->getCurrencyFormBuilder()->getFormFor($currencyId);
+            $currencyForm->handleRequest($request);
+
+            if ($currencyForm->isSubmitted()) {
```
same here.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -205,6 +251,135 @@ public function toggleStatusAction($currencyId)
+     *
+     * @return RedirectResponse
+     */
+    public function updateExchangeRatesAction()
```
action says it "updates" while command says it "refreshes", i think you choose one.

LouiseBonnard commented on src/Adapter/Currency/CommandHandler/ToggleExchangeRateAutomatizationHandler.php:
```diff
@@ -0,0 +1,262 @@
+        $isCronAdded = $cronJobsModule->addOneShotTask(
+            $cronUrl,
+            $this->translator->trans(
+                'Live exchange Rate for %shop_name%',
```
Is the capital letter here necessary? I know it was the way we used to write this wording but I would prefer _Live exchange rate for %shop_name%_ instead, in case it does not change anything. ;-)

tomas862 replied:
@LouiseBonnard The translation gets missing in `cronjob` module description row if I change the name

![missing_cron](https://user-images.githubusercontent.com/17051250/51805083-c697d980-2271-11e9-9f30-1dbf4219d3c3.png)

should I proceed with changing that anyway? :)

LouiseBonnard replied:
It's a pity... thanks for letting me know about, let's leave the string as is then. :-)

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/Improve/International/CurrencyController.php:
```diff
@@ -205,6 +251,135 @@ public function toggleStatusAction($currencyId)
+            return $this->json([
+                    'status' => false,
+                    'message' => $this->trans(
+                        'You do not have permission to update this.',
```
And there it should be _You need permission to update this._, if we apply the #12235 recommendation.

tomas862 replied:
@LouiseBonnard should I apply it now? :)

LouiseBonnard replied:
Yes, please. ;-)

LouiseBonnard replied:
Yes, thank you.

tomas862 committed:
disables scheduler class by returning empty object for further usage, adds explanation above the class

sarjon committed:
Update src/Adapter/Currency/CommandHandler/AbstractCurrencyHandler.php

Co-Authored-By: tomas862 <tilginis@gmail.com>

sarjon committed:
Update src/Adapter/Currency/CommandHandler/EditCurrencyHandler.php

Co-Authored-By: tomas862 <tilginis@gmail.com>

tomas862 committed:
some renaming and adds exit conditions on if statements

tomas862 committed:
changed some namespaces, classes names, fixed the problem with currency exchange rate is not being accepted as integer value

tomas862 committed:
fixed the problem with currency exchange rate is not being accepted as integer value

tomas862 committed:
prevents from removing default currency and disabling it in edit form when action is performed in another shops context

tomas862 committed:
prevents from disabling default currency when you are in another shop in multishop context

tomas862 committed:
applies php cs fixer

tomas862 committed:
applies new form design rendering, removes some required fields, adds extra valdidation for shop association block

tomas862 commented:
Hi @LouiseBonnard  - I also found one more use case where I needed one more translation. Very similat to the previous one. It sounds like:

`%currency% is the default currency for shop %shop_name%, and therefore cannot be disabled`.

What do you think about this translation? :)

tomas862 commented:
Pr status: waiting for answers for https://github.com/PrestaShop/PrestaShop/pull/11894#issuecomment-458115016 , https://github.com/PrestaShop/PrestaShop/pull/11894#discussion_r251248975 and https://github.com/PrestaShop/PrestaShop/pull/11894#discussion_r251248963

LouiseBonnard commented:
@tomas862, perfect, thanks! Please localize it in _Admin.International.Notification_ or _Admin.Notifications.Error_.

tomas862 committed:
changes permissions messages

tomas862 commented:
@LouiseBonnard  wording done, thanks!

matks commented:
@LouiseBonnard do you want to re-check wordings as the PR is approved and Tomas applied your feedbacks ? Else I put the "wording ✔️" label

marionf commented:
@tomas862 

1) When I click on help, it opens it like this

![capture d ecran_963](https://user-images.githubusercontent.com/13449658/52202097-c4371000-286d-11e9-81cc-0c63a369c1dc.png)

Instead of this:

![capture d ecran_964](https://user-images.githubusercontent.com/13449658/52202105-cf8a3b80-286d-11e9-89fb-f53522628716.png)

2) In a mulstishop context, if I select "all shops" or the first shop (here "prestashop") the real-time exchange rate is not saved despite the confirmation message

https://drive.google.com/open?id=1Kf0pZrR7CPHlHOHUkAeGPOJ9A3wcPsSm


@TristanLDD Can you verify if this page is ok for you ?

![capture d ecran_965](https://user-images.githubusercontent.com/13449658/52202320-78389b00-286e-11e9-89ea-7a8f19bf10d5.png)

tomas862 committed:
adds help message sidebar popup

tomas862 commented:
@marionf regarding the https://github.com/PrestaShop/PrestaShop/pull/11894#issuecomment-460194934

1. Fixed
2. Its known issue mentioned here https://github.com/PrestaShop/PrestaShop/issues/12025 . Previously it was a same problem just no confirmation message were dispalyed. Although I see the need to fix it in one way or another. I'll discuss with @matks about and let you know if we are going to fix that.

@matks I have one question related with https://github.com/PrestaShop/PrestaShop/issues/12025:

1. As described in the https://github.com/PrestaShop/PrestaShop/pull/11894#issuecomment-460194934  second point the setting does not work correctly on all shop context. Possible solution would be:

- On all shop context, I create cronjob link for each shop
- the setting is determined to be **on** only when all links are active for each shop when client is in all shop context

matks commented:
@tomas862 let's continue this discussion in https://github.com/PrestaShop/PrestaShop/issues/12025 as this PR has lived looooong enough ^^

Can you please fix the lint so we get Travis approval too 😄 ?

tomas862 committed:
applies php cs fixer

matks commented:
Thank you @tomas862 for the big work 🎉

tomas862 commented:
Thank you all for reviews!

matks commented on src/PrestaShopBundle/Resources/views/Admin/Improve/International/Currency/Blocks/form.html.twig:
```diff
@@ -0,0 +1,78 @@
+        }) }}
+
+        {% set exchangeRateLabel %}
+          {{ 'Exchange rates are calculated from one unit of your shop\'s default currency. For example, if the default currency is euros and your chosen currency is dollars, type "1.20" (1&euro; = $1.20).'|trans({}, 'Admin.International.Help') }}
```
@tomas862 Pablo notified me about this wording as `1&euro;` is being printed "as is" :
![capture d ecran 2019-02-12 a 17 51 21](https://user-images.githubusercontent.com/3830050/52653390-2c31ca00-2ef0-11e9-9193-77deb67152b1.png)

I cant find it in legacy page. Did you create it ?

tomas862 replied:
Yes it existed previously on `AdminCurrenciesController.php` line 119 - this file is currently removed due to whole page is migrated. I agree this looks strange - maybe and update here required for the transaltion?
